### PR TITLE
Added support for standalone plugin

### PR DESCRIPTION
**Pull request by aarontay**
--Created c++ flags class 
--Added command line options for standalone support
--Updated file extension
--Added command line flags to install with #include files
--Updated indentations and formatting to use 4 white spaces
--Updated formating and indentations to use 4 white spaces
--Added HeartbeatWatch support
--Added Flags class to source
--updated flags
--backing up work to rebuild
--Adding current devs to repo backup
--Adding cppnetlib support
--Checking in updated flags class and added http libs

Fixes # Add support for standalone plugin in cpp  

Summary of changes:
- Adds matching flags for c++ plugin
- Matches defaults of plugin-lib-go
- Matches plugin starting behavior for plugin-lib-go

How to verify it:
-compile example collector 'rando' to test added standalone capability

Testing done:
-verified successful build in Travis tests

A picture of a snapping turtle (not required but encouraged):
-

### CODE REVIEW

File: src/snap/flags.cc [hunk 1]
```diff
@@ -0,0 +1,395 @@
+#include "flags.h"
+//#include <snap/flags.h>
+
+#include <map>
+#include <iostream>
+#include <boost/any.hpp>
+#include <spdlog/spdlog.h>
+
+// Set default option values
+#define STAND_ALONE_PORT 8181
```
Comment (katarzyna-z) [inline src/snap/flags.cc]: It seems that `STAND_ALONE_PORT` should be changed to 8182
for plugin written in golang I see following help:
```
--stand-alone-port value      specify http port when stand-alone is set (default: 8182)

```

File: src/snap/flags.cc [hunk 2]
```diff
@@ -0,0 +1,395 @@
+#include "flags.h"
+//#include <snap/flags.h>
```
Comment (tomaszkapela) [inline src/snap/flags.cc]: Remove if no longer necessary.

File: src/snap/metric.h [hunk 3]
```diff
@@ -26,15 +26,6 @@ limitations under the License.
 
 namespace Plugin {
 
-/**
```
Comment (katarzyna-z) [inline src/snap/metric.h]: This part should not be removed.

File: src/snap/grpc_export.cc [hunk 4]
```diff
@@ -23,8 +23,8 @@ limitations under the License.
 #include <string>
 
 #include <grpc++/grpc++.h>
-
 #include <json.hpp>
+#include <boost/network/protocol/http/server.hpp>
```
Comment (katarzyna-z) [inline src/snap/grpc_export.cc]: Please update readme of snap-plugin-lib-cpp and add boost as one of dependencies, please remember to add information about boost version.

**Comment by katarzyna-z** [comment]
for now trying to build I see following error:
```
snap/grpc_export.cc:58:35: error: ‘boost::network::http::http_server<Plugin::GRPCExportImpl::handler_type>::response’ has not been declared
                      http_server::response &response) {
                                   ^
snap/grpc_export.cc: In member function ‘void Plugin::GRPCExportImpl::handler_type::operator()(const request&, int&)’:
snap/grpc_export.cc:59:33: error: ‘boost::network::http::http_server<Plugin::GRPCExportImpl::handler_type>::response’ has not been declared
         response = http_server::response::stock_reply(
                                 ^
snap/grpc_export.cc:60:26: error: ‘boost::network::http::http_server<Plugin::GRPCExportImpl::handler_type>::response’ has not been declared
             http_server::response::ok, this->preamble);

```

File: README.md [hunk 5]
```diff
@@ -52,6 +52,8 @@ Dependencies:
 * grpc++ (see below)
 * libtool
 * make
+* [boost c++ libs](http://www.boost.org)
```
Comment (katarzyna-z) [inline README.md]: Please specify which versions of these libraries should be used.

File: examples/collector/src/Makefile [hunk 6]
```diff
@@ -27,7 +27,8 @@ CC=g++
 IDIRS=-I.
 LDIRS=-L.
 CFLAGS=-std=c++1y
-LIBS=-Wl,--start-group $(SNAPLIBS)/libsnap.a $(SNAPLIBS)/libgrpc.a $(SNAPLIBS)/libgrpc++.a $(SNAPLIBS)/libprotobuf.a -Wl,--end-group -lz -lpthread
+#LIBS= -lsnap -lgrpc++ -lprotobuf -lz -lpthread -lboost_program_options
```
Comment (katarzyna-z) [inline examples/collector/src/Makefile]: Please remove commented line

File: src/snap/flags.cc [hunk 7]
```diff
@@ -0,0 +1,408 @@
...
+namespace po = boost::program_options;
+namespace spd = spdlog;
+// Group of options allowed only on command line
+int Plugin::Flags::addDefaultCommandFlags() {
+    try {
+        _command.add_options()
+            ("help,h", "Display available command-line options")
+            ("version,v", "Print the version");
+
+
```
Comment (katarzyna-z) [inline src/snap/flags.cc]: unnecessary empty lines

File: examples/collector/src/rando.cc [hunk 8]
```diff
@@ -41,172 +47,180 @@ enum supported_types
...
-  Rando plg = Rando();
-  start_collector(&plg, meta);
+int main(int argc, char **argv) {
+    Flags cli;
+    cli.SetFlags();
+    cli.ParseFlags(argc, argv);
+    
+    Meta meta(Type::Collector, "rando", 1);
+
+    if (cli.GetFlagsVM().count("version")) {
```
Comment (zachary-trudo) [inline examples/collector/src/rando.cc]: Move this section after the flags parsing - no need to create meta  object if we are just going to exit.

File: examples/collector/src/Makefile [hunk 9]
```diff
@@ -27,7 +27,7 @@ CC=g++
 IDIRS=-I.
 LDIRS=-L.
 CFLAGS=-std=c++1y
-LIBS=-Wl,--start-group $(SNAPLIBS)/libsnap.a $(SNAPLIBS)/libgrpc.a $(SNAPLIBS)/libgrpc++.a $(SNAPLIBS)/libprotobuf.a -Wl,--end-group -lz -lpthread
+LIBS=-Wl,--start-group $(SNAPLIBS)/libsnap.a $(SNAPLIBS)/libgrpc.a $(SNAPLIBS)/libgrpc++.a $(SNAPLIBS)/libprotobuf.a -Wl,--end-group -lboost_program_options -lboost_system -lboost_thread -lcppnetlib-uri -lcppnetlib-client-connections -lcppnetlib-server-parsers -lz -lpthread -lssl -lcrypto
```
Comment (mkleina) [inline examples/collector/src/Makefile]: Please add this change to all plugins. Running `make` in examples folder fails otherwise.

**Comment by mkleina** [comment]
@aarontay This PR contains stand-alone only for collectors? Processor and publisher from examples does not start in standalone mode:
```json
$ ./test-processor-graffiti --stand-alone
{"ErrMessage":"","ListenAddress":"127.0.0.1:44877","Meta":{"CacheTTL":500,"ConcurrencyCount":5,"Exclusive":false,"Name":"graffiti","RPCType":2,"RPCVersion":1,"RoutingStrategy":0,"StandAloneEnabled":false,"StandAlonePort":0,"Type":1,"Unsecure":true,"Version":1},"State":0,"Type":1,"Version":1}
```

As you see, output contains "StandAlonePort":0, using --stand-alone-port does not change anything, same response.

File: examples/collector/src/rando.cc [hunk 10]
```diff
@@ -41,172 +47,180 @@ enum supported_types
...
-int main() {
-  Meta meta(Type::Collector, "rando", 1);
-  Rando plg = Rando();
-  start_collector(&plg, meta);
+int main(int argc, char **argv) {
+    Flags cli;
+    cli.SetFlags();
+    cli.ParseFlags(argc, argv);
+    
+    Meta meta(Type::Collector, "rando", 1);
```
Comment (aarontay) [inline examples/collector/src/rando.cc]: done

File: examples/collector/src/Makefile (hunk 9 above)
Comment (aarontay) [inline examples/collector/src/Makefile]: Updated all plugins to now include flags and have stand-alone option

File: src/snap/grpc_export.cc [hunk 11]
```diff
@@ -48,81 +48,146 @@ using json = nlohmann::json;
...
-  }
-  builder.reset(new grpc::ServerBuilder());
-  builder->AddListeningPort(server_address, grpc::InsecureServerCredentials(),
-                           &this->port);
-
+    std::stringstream ss;
+    ss << this->meta->listen_addr << ":";
+    this->meta->listen_port == "" ? ss << "0" : ss << this->meta->listen_port;
+
+    //string server_address = "127.0.0.1:0";
```
Comment (katarzyna-z) [inline src/snap/grpc_export.cc]: please remove commented code

File: src/snap/grpc_export.cc [hunk 12]
```diff
@@ -48,81 +48,146 @@ using json = nlohmann::json;
...
+        case Plugin::Collector:
+            this->service.reset(new Proxy::CollectorImpl(plugin->IsCollector()));
+            break;
+        case Plugin::Processor:
+            this->service.reset(new Proxy::ProcessorImpl(plugin->IsProcessor()));
+            break;
+        case Plugin::Publisher:
+            this->service.reset(new Proxy::PublisherImpl(plugin->IsPublisher()));
+            break;
+        //TODO: aarontay
```
Comment (katarzyna-z) [inline src/snap/grpc_export.cc]: I suggest to remove this section of commented code

File: src/snap/grpc_export.cc [hunk 13]
```diff
@@ -48,81 +48,146 @@ using json = nlohmann::json;
...
+        //case Plugin::StreamCollector
+        //
+        // break;
+        //default:
```
Comment (katarzyna-z) [inline src/snap/grpc_export.cc]: default case could be useful, please consider to uncomment this part or remove it

File: src/snap/plugin.cc [hunk 14]
```diff
@@ -38,77 +38,112 @@ using grpc::Server;
...
-  auto plugin_ptr = shared_ptr<Plugin::PluginInterface>(plugin, [](void*){});
-  auto completion = exporter->ExportPlugin(plugin_ptr, &meta);
-  completion.get();
-}
+    auto exporter = Plugin::LibSetup::exporter_provider();
+    // disable deleting the plugin instance
+    auto plugin_ptr = shared_ptr<Plugin::PluginInterface>(plugin, [](void*){});
+    auto completion = exporter->ExportPlugin(plugin_ptr, &meta);
+    completion.get();
+}
```
Comment (katarzyna-z) [inline src/snap/plugin.cc]: Please add a new line at the end of this file.

File: src/snap/plugin.cc [hunk 15]
```diff
@@ -38,77 +38,112 @@ using grpc::Server;
 using grpc::ServerBuilder;
 
 static void start_plugin(Plugin::PluginInterface* plugin, const Plugin::Meta& meta);
+int start_stand_alone(Plugin::PluginInterface* plugin, const Plugin::Meta& meta);
 
 function<unique_ptr<Plugin::PluginExporter, function<void(Plugin::PluginExporter*)>>()> Plugin::LibSetup::exporter_provider = []{ return std::unique_ptr<PluginExporter>(new GRPCExporter()); };
 
-Plugin::PluginException::PluginException(const string& message) :
-                                         runtime_error(message) {}
+Plugin::PluginException::PluginException(const std::string& message) :
```
Comment (katarzyna-z) [inline src/snap/plugin.cc]: unneeded space character at the end of this line

**Comment by katarzyna-z** [comment]
I'v approved this pull request. but LGTM after small clearance - I added several comments, please take a look at them.

File: src/snap/grpc_export.cc (hunk 13 above)
Comment (aarontay) [inline src/snap/grpc_export.cc]: Removed from this PR; I have theses implemented in the upcoming streaming support already

File: src/snap/config.cc [hunk 16]
```diff
@@ -27,62 +27,62 @@ ConfigPolicy::ConfigPolicy() {}
...
-  auto int_map = rpc_map.intmap();
-  return int_map.at(key);
+    auto int_map = rpc_map.intmap();
+    return int_map.at(key);
 }
 
 std::string Config::get_string(const std::string& key) const {
-  auto str_map = rpc_map.stringmap();
-  return str_map.at(key);
+    auto str_map = rpc_map.stringmap();
```
Comment (aarontay) [inline src/snap/config.cc]: Thanks for the feedback! These changes fall outside of the scope of this PR, so I created [issue 49](https://github.com/intelsdi-x/snap-plugin-lib-cpp/issues/49] to ensure we address them.

**Pull request merged**